File: android/app/src/main/java/de/WyLight/WyLight/library/Endpoint.java
/**
                Copyright (C) 2012 - 2014 Patrick Bruenn.

    This file is part of WyLight.

    Wylight is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    WiflyLight is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with WiflyLight.  If not, see <http://www.gnu.org/licenses/>. */

package de.WyLight.WyLight.library;

import java.io.Serializable;

import de.WyLight.WyLight.exception.FatalError;

public class Endpoint implements Serializable {
	private static final long serialVersionUID = -8653676704955283379L;

	private native long connect(long pBroadcastReceiver, long fingerprint)
			throws FatalError;

	private native String getEndpointName(long pBroadcastReceiver,
			long fingerprint);

	private native void setEndpointName(long pBroadcastReceiver,
			long fingerprint, String deviceId) throws FatalError;

	private final int mAddr;
	private final short mPort;
	private final long mParentBroadcastReceiver;
	private boolean mIsOnline = false;

	public Endpoint(long parent, int addr, short port) {
		mAddr = addr;
		mPort = port;
		mParentBroadcastReceiver = parent;
	}

	public Endpoint(long parent, long fingerprint) {
		this(parent, (int) (fingerprint >> 32),
				(short) (fingerprint & 0x000000000000ffffL));
	}

	public long connect() throws FatalError {
		return connect(mParentBroadcastReceiver, getFingerprint());
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (obj == null)
			return false;
		if (getClass() != obj.getClass())
			return false;
		Endpoint other = (Endpoint) obj;
		if (mAddr != other.mAddr)
			return false;
		if (mParentBroadcastReceiver != other.mParentBroadcastReceiver)
			return false;
		if (mPort != other.mPort)
			return false;
		return true;
	}

	public int getAddr() {
		return mAddr;
	}

	public long getFingerprint() {
		return (((long) mAddr) << 32) | ((long) mPort);
	}

	/**
	 * Get ip address string
	 * @return aaa.aaa.aaa.aaa
	 */
	public String getAddress() {
		int _4 = mAddr & 0xff;
		int _3 = Integer.rotateRight(mAddr, 8) & 0xff;
		int _2 = Integer.rotateRight(mAddr, 16) & 0xff;
		int _1 = Integer.rotateRight(mAddr, 24) & 0xff;
		return Integer.toString(_1) + '.' + Integer.toString(_2) + '.'
				+ Integer.toString(_3) + '.' + Integer.toString(_4);
	}

	/**
	 * Get ip and port as a combined string
	 * @return aaa.aaa.aaa.aaa:ppppp
	 */
	public String getAddressPort() {
		return getAddress() + ':' + String.valueOf(mPort);
	}

	public String getName() {
		return getEndpointName(mParentBroadcastReceiver, getFingerprint());
	}

	public short getPort() {
		return mPort;
	}

	@Override
	public int hashCode() {
		final int prime = 31;
		int result = 1;
		result = prime * result + mAddr;
		result = prime
				* result
				+ (int) (mParentBroadcastReceiver ^ (mParentBroadcastReceiver >>> 32));
		result = prime * result + mPort;
		return result;
	}

	public boolean isOnline() {
		return mIsOnline;
	}

	public boolean isValid() {
		return (0 != mAddr) && (0 != mPort);
	}

	public void setName(String deviceId) throws FatalError {
		setEndpointName(mParentBroadcastReceiver, getFingerprint(), deviceId);
	}

	public void setOnline(boolean isOnline) {
		mIsOnline = isOnline;
	}

	@Override
	public String toString() {
		return getName();
	}
}

File: library/Endpoint.cpp
/**
                Copyright (C) 2013 - 2015 Nils Weiss, Patrick Bruenn.

    This file is part of WyLight.

    WyLight is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    WyLight is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with WyLight.  If not, see <http://www.gnu.org/licenses/>. */

#include "Endpoint.h"

namespace WyLight
{
Endpoint::Endpoint(const BroadcastMessage& msg, sockaddr_in* addr)
    : mIp(ntohl(addr->sin_addr.s_addr)),
    mPort(ntohs(msg.port)),
    mScore(1),
    mDeviceId(&msg.deviceId[0],
              strnlen(&msg.deviceId[0],
                      sizeof(msg.deviceId))),
    mType(msg.IsRN171Broadcast() ? RN171 : CC3200)
{}

Endpoint::Endpoint(uint32_t            ip,
                   uint16_t            port,
                   uint8_t             score,
                   std::string         devId,
                   enum Endpoint::TYPE type)
    : mIp(ip),
    mPort(port),
    mScore(score),
    mDeviceId(devId),
    mType(type)
{}

bool Endpoint::operator<(const Endpoint& ref) const
{
    return (mIp < ref.GetIp())
           || ((mIp == ref.GetIp()) && (mPort < ref.GetPort()));
}

void Endpoint::WriteTo(std::ostream& out) const
{
    out << (int)mScore << ' ' <<
        std::hex << GetIp() << ' ' <<
        std::dec << GetPort() << ' ' <<
        GetDeviceId() << '\n';
}

std::ostream& operator<<(std::ostream& out, const Endpoint& ref)
{
    return out << (int)ref.mScore << ' ' <<
           ((ref.mIp & 0xff000000) >> 24) << '.' <<
           ((ref.mIp & 0x00ff0000) >> 16) << '.' <<
           ((ref.mIp & 0x0000ff00) >> 8) << '.' <<
           (ref.mIp & 0x000000ff) <<
           ':' << ref.mPort <<
           "  :  " << ref.mDeviceId;
}

std::istream& operator>>(std::istream& in, Endpoint& ref)
{
    std::string ip;
    in >> ref.mScore >> ip >> ref.mPort >> ref.mDeviceId;
    ref.mIp = WiflyColor::ToARGB(ip);
    return in;
}

bool operator==(const Endpoint& lhs, const Endpoint& rhs)
{
    if ((lhs.mDeviceId == rhs.mDeviceId) &&
        (lhs.mIp == rhs.mIp) &&
        (lhs.mPort == rhs.mPort) &&
        (lhs.mScore == rhs.mScore)
        )
        return true;
    else
        return false;
}

bool operator!=(const Endpoint& lhs, const Endpoint& rhs)
{
    return !(lhs == rhs);
}

uint64_t Endpoint::AsUint64(void) const
{
    return ((uint64_t)mIp << 32) | mPort;
}

std::string Endpoint::GetDeviceId(void) const
{
    return mDeviceId;
}

uint32_t Endpoint::GetIp(void) const
{
    return mIp;
}

uint16_t Endpoint::GetPort(void) const
{
    return mPort;
}

uint8_t Endpoint::GetScore(void) const
{
    return mScore;
}

enum Endpoint::TYPE Endpoint::GetType(void) const
{
    return mType;
}

void Endpoint::SetDeviceId(const std::string& deviceId)
{
    mDeviceId = deviceId;
}

void Endpoint::SetScore(const uint8_t score)
{
    mScore = score;
}

Endpoint& Endpoint::operator++(void)
{
    if (mScore < 255) ++mScore;
    return *this;
}

bool Endpoint::IsValid(void) const
{
    return (0 != mIp) && (0 != mPort);
}
}
